crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -150,42 +154,30 @@ where
         input: &Message,
     ) -> Result<Vec<Message>, ConversionError> {
         let tedge_event = ThinEdgeEvent::try_from(input.topic.name.as_str(), input.payload_str()?)?;
+        let c8y_event = C8yCreateEvent::try_from(tedge_event)?;
 
         // If the message size is well within the Cumulocity MQTT size limit, use MQTT to send the mapped event as well
         if input.payload_bytes().len() < self.size_threshold.0 {
-            let smartrest_alarm = event::serialize_event(tedge_event)?;
-            let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
+            if c8y_event.extras.is_empty() {
+                let smartrest_event = format!(
+                    "{},{},\"{}\",{}",
+                    CREATE_EVENT_SMARTREST_CODE,
+                    c8y_event.event_type,
+                    c8y_event.text,
+                    c8y_event.time
+                );
+                let smartrest_topic = Topic::new_unchecked(SMARTREST_PUBLISH_TOPIC);
 
-            Ok(vec![Message::new(&smartrest_topic, smartrest_alarm)])
+                Ok(vec![Message::new(&smartrest_topic, smartrest_event)])
+            } else {
+                let cumulocity_event_json = serde_json::to_string(&c8y_event)?;
+                let json_mqtt_topic = Topic::new_unchecked(C8Y_JSON_MQTT_EVENTS_TOPIC);
+
+                Ok(vec![Message::new(&json_mqtt_topic, cumulocity_event_json)])
+            }
         } else {
             // If the message size is larger than the MQTT size limit, use HTTP to send the mapped event
-            let (event_text, event_time) = match tedge_event.data {
-                None => {
-                    let message = tedge_event.name.clone();
-                    let time = OffsetDateTime::now_utc().format(&Rfc3339)?;
-
-                    (message, time)
-                }
-                Some(event_data) => {
-                    let message = event_data
-                        .message
-                        .unwrap_or_else(|| tedge_event.name.clone());
-                    let time = event_data.time.map_or_else(
-                        || OffsetDateTime::now_utc().format(&Rfc3339),
-                        |timestamp| timestamp.format(&Rfc3339),
-                    )?;
-                    (message, time)
-                }
-            };
-
-            let _ = self
-                .http_proxy
-                .send_event(
-                    tedge_event.name.as_str(),
-                    event_text.as_str(),
-                    Some(event_time),
-                )
-                .await?;
+            let _ = self.http_proxy.send_event(c8y_event).await?;
```

Comment:
So If I understand correctly we have 3 cases:
1) large message => sent over HTTP serialized using c8y json
2) small message with extras => sent over MQTT using c8y json
3) small & simple event => sent over MQTT using smartrest.

Am I correct?

I think that things can be made clearer. For instance, I'm surprised that the threshold is on the received message size, not the sent one.

Reply:
> Am I correct?

Yes, that's correct.

> I think that things can be made clearer.

Will add more inline documentation.

> I'm surprised that the threshold is on the received message size, not the sent one.

Good catch. I'll fix that logic to do that switching based on mapped message size rather than the received message size.